#!/bin/R
### first run combine pops for multi-pop cross objects

pop <- 'ELR'

source("/home/jmiller1/QTL_agri/MAP/control_file.R")
library("ggridges")
library("plyr")
library("scales")
library("ggrepel")
library('qtl')
library('RColorBrewer')

mpath <- '/home/jmiller1/QTL_agri/data'
setwd(mpath)

##keeping colors consistent####################
all.pops <- c("NBH", "BRP", "ELR", "NEW")
popcol <- brewer.pal(8, "Paired")[c(2, 4, 6, 8)]
names(popcol) <- all.pops

popgen <- popcol
names(popgen) <- c('NBH','BP','ER','NYC')

popout <- c(popgen,'grey')
names(popout) <- c('NBH','BP','ER','NYC','BI')

### Color for stat comparisons
statcol <- popcol
names(statcol) <- c('BI.NBH','ER.KC','BP.F','NYC.SH')
################################################

pbs <- file.path(mpath, 'pbstat.txt.ncbi.lifted')
pbs <- read.table(pbs, sep = "\t", header = T)
pbs$mid <- pbs$V2 + (abs(pbs$V3 - pbs$V2) * .5)
pbs$V1 <- gsub('chr',"",pbs$V1)

pfst <- file.path(mpath, 'pfst.txt.ncbi.lifted')
pfst <- read.table(pfst, sep = "\t", header = T)
pfst$mid <- pfst$start + (abs(pfst$end - pfst$start) * .5)
pfst$Scaffold <- gsub('chr',"",pfst$Scaffold)

taj <- file.path(mpath, 'tajstat.txt.ncbi.lifted')
taj <- read.table(taj, sep = "\t", header = T)
taj$mid <- taj$start + (abs(taj$end - taj$start) * .5)
taj$Scaffold <- gsub('chr',"",taj$Scaffold)

pi <- file.path(mpath, 'piper.txt.ncbi.lifted')
pi <- read.table(pi, sep = "\t", header = T)
pi$mid <- pi$start + (abs(pi$end - pi$start) * .5)
pi$Scaffold <- gsub('chr',"",pi$Scaffold)


plot_stat <- function(Z,ch,poplot){

  ind <- which(Z[,1] == ch)

  pops <- names(poplot)

  ymx_mn <- c(
    quantile(as.matrix(Z[ind,pops]), probs = 0.00001, na.rm = T),
    quantile(as.matrix(Z[ind,pops]), probs = 0.99999, na.rm = T))

  x_mx_mn <- c(min(Z[ind,'mid'],na.rm=T),max(Z[ind,'mid'],na.rm=T))

  X <- Z[ind,'mid']

  Y <- as.list(Z[ind,pops])
  names(Y) <- pops

  plot(x_mx_mn, ymx_mn, type="n")
  sapply(pops,plot_pnts,X,Y,poplot)

}

plot_pnts <- function(stat,X,Y,poplot){ points(X, Y[[stat]], pch=20, col=poplot[stat]) }


png("/home/jmiller1/public_html/pfst.png", width = 3000)
plot_stat(pfst,ch=2,poplot=statcol)
dev.off()

png("/home/jmiller1/public_html/pbs.png", width = 3000)
plot_stat(pbs,ch=2,poplot=popgen)
dev.off()

png("/home/jmiller1/public_html/taj.png", width = 3000)
plot_stat(taj,ch=2,poplot=popout)
dev.off()

plot_stat_sep <- function(Z,ch,poplot){

  ind <- which(Z[,1] == ch)

  pops <- names(poplot)

  ymx_mn <- c(
    quantile(as.matrix(Z[ind,pops]), probs = 0.00001, na.rm = T),
    quantile(as.matrix(Z[ind,pops]), probs = 0.99999, na.rm = T))

  x_mx_mn <- c(min(Z[ind,'mid'],na.rm=T),max(Z[ind,'mid'],na.rm=T))

  X <- Z[ind,'mid']

  Y <- as.list(Z[ind,pops])
  names(Y) <- pops

  par(mfrow=c(length(pops),1),mar = c(1, 1, 1, 1),oma = c(1.5, 1.5, 1.5, 1.5))

  sapply(pops,plot_pop_sep,X,Y,poplot,x_mx_mn,ymx_mn)

  axis(side=1)

}

plot_pop_sep <- function(stat,X,Y,poplot,x_mx_mn,ymx_mn){
 plot(x_mx_mn, ymx_mn, type="n",xaxs="i", yaxs="i",main=NULL,xaxt="n",bty='n')
 points(X, Y[[stat]], pch=20, col=poplot[stat])
}

png("/home/jmiller1/public_html/pfst.png", width = 1000)
plot_stat_sep(pfst,ch=18,poplot=statcol)
dev.off()



#get_popgen <- function(X){
# ind <- which.min(abs(popgen[which(popgen[,'V1'] == X[1] ),'mid'] - X[3]))
# popgen[which(popgen[,'V1'] == X[1] ),][ind,]
#}

#### AHRs #####
AHR.bed <- read.table("lift_AHR_genes.bed", stringsAsFactors = F, header = F)
colnames(AHR.bed) <- c("chrom", "str", "stp", "gene")
AHR.bed$chrom <- as.numeric(gsub("chr", "", AHR.bed$chrom))
AHR.bed$str <- as.numeric(AHR.bed$str)
AHR.bed$stp <- as.numeric(AHR.bed$stp)
AHR.notmap <- AHR.bed[is.na(AHR.bed$chrom), ]
AHR.bed <- AHR.bed[!is.na(AHR.bed$chrom), ]
AHR.bed$gene <- gsub(":158640", "", AHR.bed$gene)
# add arnts (forgot to scan for them)
################################################

## Phenotypes
################################################
cross.BRP <- read.cross(format = "csv", dir = mpath, file = 'brp.mapped.tsp.csv', genotypes=c("1","2","3"), estimate.map = FALSE)
cross.ELR <- read.cross(format = "csv", dir = mpath, file = 'ELR.mapped.tsp.csv', genotypes=c("1","2","3"), estimate.map = FALSE)
cross.NBH <- read.cross(format = "csv", dir = mpath, file = 'NBH.mapped.tsp.csv', genotypes=c("1","2","3"), estimate.map = FALSE)
cross.NEW <- read.cross(format = "csv", dir = mpath, file = 'NEW.mapped.tsp.csv', genotypes=c("1","2","3"), estimate.map = FALSE)
################################################

################################################
get_cor <- function(Z){
 mp <- pull.map(Z)
 pos <- lapply(mp,chr_names_pos)
 mapply(cor,mp,pos)
}
chr_names_pos <- function(X){
 b <- as.numeric(gsub("*.:",'',names(X)))
 ifelse(is.na(b),0,b)
}
################################################

################################################
cor_nbh <- get_cor(cross.NBH)
cor_elr <- get_cor(cross.ELR)
cor_brp <- get_cor(cross.BRP)
cor_new <- get_cor(cross.NEW)

cross.BRP <- flip.order(cross.BRP, names(cor_brp)[which(cor_brp < 0)])
cross.NBH <- flip.order(cross.NBH, names(cor_nbh)[which(cor_nbh < 0)])
cross.NEW <- flip.order(cross.NEW, names(cor_new)[which(cor_new < 0)])
cross.ELR <- flip.order(cross.ELR, names(cor_elr)[which(cor_elr < 0)])
################################################

################################################
cross.nbh <- sim.geno(cross.NBH, n.draws = 500, step = 5, off.end = 10, error.prob = 0.025,
  map.function = "kosambi", stepwidth = "fixed")
cross.new <- sim.geno(cross.NEW, n.draws = 500, step = 5, off.end = 10, error.prob = 0.025,
  map.function = "kosambi", stepwidth = "fixed")
cross.elr <- sim.geno(cross.ELR, n.draws = 500, step = 5, off.end = 10, error.prob = 0.025,
  map.function = "kosambi", stepwidth = "fixed")
cross.brp <- sim.geno(cross.BRP, n.draws = 500, step = 5, off.end = 10, error.prob = 0.025,
  map.function = "kosambi", stepwidth = "fixed")
################################################

################################################
cross.nbh <- reduce2grid(cross.nbh)
cross.new <- reduce2grid(cross.new)
cross.elr <- reduce2grid(cross.elr)
cross.brp <- reduce2grid(cross.brp)
################################################

################################################
scan.norm.imp.NBH <- scanone(cross.nbh, method = "imp", model = "normal", pheno.col = 5)
scan.bin.imp.NBH <-  scanone(cross.nbh, method = "em", model = "binary", pheno.col = 4)
scan.norm.imp.ELR <- scanone(cross.elr, method = "imp", model = "normal", pheno.col = 5)
scan.bin.imp.ELR <-  scanone(cross.elr, method = "em", model = "binary", pheno.col = 4)
scan.norm.imp.NEW <- scanone(cross.new, method = "imp", model = "normal", pheno.col = 5)
scan.bin.imp.NEW <-  scanone(cross.new, method = "em", model = "binary", pheno.col = 4)
scan.norm.imp.BRP <- scanone(cross.brp, method = "imp", model = "normal", pheno.col = 5)
scan.bin.imp.BRP <-  scanone(cross.brp, method = "em", model = "binary", pheno.col = 4)
################################################

################################################
### use scanone for plots
themelt.nbh <- scan.bin.imp.NBH
themelt.new <- scan.bin.imp.NEW
themelt.elr <- scan.bin.imp.ELR
themelt.brp <- scan.bin.imp.BRP

themelt.nbh$pop <- "NBH"
themelt.new$pop <- "NEW"
themelt.elr$pop <- "ELR"
themelt.brp$pop <- "BRP"

save.image('08_phys_plots_pos.rsave')
################################################

################################################
### get positions of genes
nbh.gens <- cnv.ahrs(cross.nbh, AHRdf = AHR.bed, EXP = F)
new.gens <- cnv.ahrs(cross.new, AHRdf = AHR.bed, EXP = F)
elr.gens <- cnv.ahrs(cross.elr, AHRdf = AHR.bed, EXP = F)
brp.gens <- cnv.ahrs(cross.brp, AHRdf = AHR.bed, EXP = F)

qtl.gens <- nbh.gens[which(nbh.gens$chr %in% c(1, 2, 5, 8, 10, 12, 13, 18, 24)),]
minor.gens <- nbh.gens[which(nbh.gens$chr %in% c(8, 13, 23, 24)), ]
incompat.gens <- nbh.gens[which(nbh.gens$chr %in% c(8, 13)), ]
qtl_pg <- c(2,8, 13, 18, 24)
ol.gens <- nbh.gens[which(nbh.gens$chr %in% qtl_pg), ]
################################################

################################################
### ggplot popgen locations
nbh.popgen <- read.table("outliersNBH.txt.ncbi.lifted", sep = "\t", header = T)
new.popgen <- read.table("outliersNYC.txt.ncbi.lifted", sep = "\t", header = T)
elr.popgen <- read.table("outliersER.txt.ncbi.lifted", sep = "\t", header = T)
brp.popgen <- read.table("outliersBP.txt.ncbi.lifted", sep = "\t", header = T)
################################################

################################################
### Use nbh coords but elr and new popgen
new.rank <- cnv.popgen(cross.nbh, new.popgen, top = 50)
nbh.rank <- cnv.popgen(cross.nbh, nbh.popgen, top = 50)
elr.rank <- cnv.popgen(cross.nbh, elr.popgen, top = 50)
brp.rank <- cnv.popgen(cross.nbh, brp.popgen, top = 50)

nbh.rank$pop <- "NBH"
new.rank$pop <- "NEW"
elr.rank$pop <- "ELR"
brp.rank$pop <- "BRP"

all.rank <- rbind(new.rank, nbh.rank, elr.rank, brp.rank)
all.rank$pop <- factor(all.rank$pop, levels = c("NBH", "BRP", "NEW", "ELR"))
qtl.rank <- all.rank[which(all.rank$chr %in% c(1,2,5,8,10,12,13,18,23,24)),]
minor.rank <- all.rank[which(all.rank$chr %in% c(8, 13, 23, 24)), ]
incompat.rank <- all.rank[which(all.rank$chr %in% c(8, 13)), ]

qtl_pg <- c(2,8, 13, 18, 24)
ol.rank <- all.rank[which(all.rank$chr %in% qtl_pg), ]
################################################

################################################
### GGriges plot
melted.nbh <- data.frame(pop = "NBH", chr = scan.norm.imp.NBH$chr, pos = scan.norm.imp.NBH$pos,
  lod = scan.norm.imp.NBH$lod)
melted.new <- data.frame(pop = "NEW", chr = scan.norm.imp.NEW$chr, pos = scan.norm.imp.NEW$pos,
  lod = scan.norm.imp.NEW$lod)
melted.elr <- data.frame(pop = "ELR", chr = scan.norm.imp.ELR$chr, pos = scan.norm.imp.ELR$pos,
  lod = scan.norm.imp.ELR$lod)
melted.brp <- data.frame(pop = "BRP", chr = scan.norm.imp.BRP$chr, pos = scan.norm.imp.BRP$pos,
  lod = scan.norm.imp.BRP$lod)

melted <- rbind(melted.nbh, melted.new, melted.elr, melted.brp)
melted$pop <- factor(melted$pop, levels = rev(c("NBH", "BRP", "NEW", "ELR")))

## Total CM length of NBH. Rescale to NBH
mxes <- sapply(1:24, function(X) {
  max(themelt.nbh$pos[which(themelt.nbh$chr == X)])
})

melso <- function(tomelt){
 ts <- tomelt[which(tomelt$chr == 1), ]
 ts$pos <- rescale(ts$pos, to = c(-10, mxes[1]))
 the_rescale <- ts
 for (i in 2:24) {
   ts <- tomelt[which(tomelt$chr == i), ]
   ts$pos <- rescale(ts$pos, to = c(-10, mxes[i]))
   the_rescale <- rbind(the_rescale, ts)
  }
 the_rescale
}

new.rescale <- melso(themelt.new)
brp.rescale <- melso(themelt.brp)
elr.rescale <- melso(themelt.elr)



##ts <- themelt.new[which(themelt.new$chr == 1), ]
##ts$pos <- rescale(ts$pos, to = c(-10, mxes[1]))
##new.rescale <- ts
##for (i in 2:24) {
##  ts <- themelt.new[which(themelt.new$chr == i), ]
##  ts$pos <- rescale(ts$pos, to = c(-10, mxes[i]))
##  new.rescale <- rbind(new.rescale, ts)
##}

##ts <- themelt.elr[which(themelt.elr$chr == 1), ]
##ts$pos <- rescale(ts$pos, to = c(-10, mxes[1]))
##elr.rescale <- ts
##for (i in 2:24) {
##  ts <- themelt.elr[which(themelt.elr$chr == i), ]
##  ts$pos <- rescale(ts$pos, to = c(-10, mxes[i]))
##  elr.rescale <- rbind(elr.rescale, ts)
##}
##
##ts <- themelt.brp[which(themelt.brp$chr == 1), ]
##ts$pos <- rescale(ts$pos, to = c(-10, mxes[1]))
##brp.rescale <- ts
##for (i in 2:24) {
##  ts <- themelt.brp[which(themelt.brp$chr == i), ]
##  ts$pos <- rescale(ts$pos, to = c(-10, mxes[i]))
##  brp.rescale <- rbind(brp.rescale, ts)
##}

allmelt <- rbind(themelt.nbh, new.rescale, elr.rescale, brp.rescale)
allmelt$pop <- factor(allmelt$pop, levels = c("NBH", "BRP", "NEW", "ELR"))
qtlmelt <- allmelt[which(allmelt$chr %in% c(1,2,5,8,10,12,13,18,19,23,24)),
  ]
qtlminor <- allmelt[which(allmelt$chr %in% c(8,13,19,23,24)), ]
incompat <- allmelt[which(allmelt$chr %in% c(8,13)), ]

qtl_pg <- c(2,8, 13, 18, 24)
ol.melt <- allmelt[which(allmelt$chr %in% qtl_pg), ]

#### MAP BRP to NBH before simcross
brp.remap <- cnv.premap(cross.nbh, cross.brp)
##save.image("/home/jmiller1/public_html/BRP_remap.Rsave")

save.image('/home/jmiller1/public_html/QTL_plot.Rsave')
################################################

################################################
p <- ggplot(themelt.nbh, aes(x = pos, y = lod))
png("/home/jmiller1/public_html/all_popgen_rank_scaled.qtl.png", width = 3000)
p + facet_wrap(~chr, scales = "free_x", nrow = 1, ncol = 24) + scale_y_continuous(limits = c(-12,
  23)) + geom_line(size = 2, alpha = 0.6) + theme_minimal() + theme(axis.text = element_text(size = 10)) +
  labs(x = "Chromosome", y = "LOD", linetype = "") + geom_label_repel(aes(x = pos,
  y = -0.1, label = gene), box.padding = unit(0.25, "lines"), parse = T, point.padding = unit(0.2,
  "lines"), force = 10, label.padding = unit(0.2, "lines"), ylim = c(0, -12), segment.size = 1,
  max.iter = 6000, data = nbh.gens, direction = "y", size = 4, seed = 666, nudge_y = -0.01,
  vjust = 3) + geom_label_repel(aes(x = pos, y = -0.1, label = gene), box.padding = unit(0.25,
  "lines"), parse = T, point.padding = unit(0.2, "lines"), force = 10, label.padding = unit(0.2,
  "lines"), ylim = c(0, -12), segment.size = 0, max.iter = 6000, data = nbh.gens,
  direction = "y", size = 4, seed = 666, nudge_y = -0.01, vjust = 3) + geom_point(aes(size = rank),
  data = nbh.rank) + geom_label_repel(aes(x = pos, y = lod, label = rank), data = nbh.rank,
  size = 4, box.padding = unit(0.35, "lines"), point.padding = unit(0.3, "lines"),
  vjust = 1)
dev.off()

p <- ggplot(themelt.new, aes(x = pos, y = lod))
png("/home/jmiller1/public_html/new_genes_below_rank_above.qtl.png", width = 3000)
p + facet_wrap(~chr, scales = "free_x", nrow = 1, ncol = 24) + scale_y_continuous(limits = c(-12,
  23)) + geom_line(size = 2, alpha = 0.6) + theme_minimal() + theme(axis.text = element_text(size = 10)) +
  labs(x = "Chromosome", y = "LOD", linetype = "") + geom_label_repel(aes(x = pos,
  y = -0.1, label = gene), box.padding = unit(0.25, "lines"), parse = T, point.padding = unit(0.2,
  "lines"), force = 10, label.padding = unit(0.2, "lines"), ylim = c(0, -12), segment.size = 1,
  max.iter = 6000, data = new.gens, direction = "y", size = 4, seed = 666, nudge_y = -0.01,
  vjust = 3) + geom_label_repel(aes(x = pos, y = -0.1, label = gene), box.padding = unit(0.25,
  "lines"), parse = T, point.padding = unit(0.2, "lines"), force = 10, label.padding = unit(0.2,
  "lines"), ylim = c(0, -12), segment.size = 0, max.iter = 6000, data = new.gens,
  direction = "y", size = 4, seed = 666, nudge_y = -0.01, vjust = 3) + geom_label_repel(aes(x = pos,
  y = lod, label = rank), data = nbh.rank, size = 4, box.padding = unit(0.35, "lines"),
  point.padding = unit(0.3, "lines"), vjust = 1)
dev.off()

p <- ggplot(themelt.elr, aes(x = pos, y = lod, color = pop))
png("/home/jmiller1/public_html/rank_top_scaled_random.qtl.png", width = 3000)
p + facet_wrap(~chr, scales = "free_x", nrow = 1, ncol = 24) + scale_y_continuous(limits = c(-12,
  23)) + scale_color_manual(values = popcol) + geom_line(size = 2, alpha = 0.6) +
  theme_minimal() + theme(axis.text = element_text(size = 10)) + labs(x = "Chromosome",
  y = "LOD", linetype = "") + geom_label_repel(aes(x = pos, y = -0.1, label = gene),
  box.padding = unit(0.25, "lines"), parse = T, point.padding = unit(0.2, "lines"),
  force = 10, label.padding = unit(0.2, "lines"), ylim = c(0, -12), segment.size = 1,
  max.iter = 6000, data = elr.gens, direction = "y", size = 4, seed = 666, nudge_y = -0.01,
  vjust = 3) + geom_label_repel(aes(x = pos, y = -0.1, label = gene), box.padding = unit(0.25,
  "lines"), parse = T, point.padding = unit(0.2, "lines"), force = 10, label.padding = unit(0.2,
  "lines"), ylim = c(0, -12), segment.size = 0, max.iter = 6000, data = elr.gens,
  direction = "y", size = 4, seed = 666, nudge_y = -0.01, vjust = 3) + geom_label_repel(aes(x = pos,
  y = lod, label = rank), data = elr.rank, size = 4, box.padding = unit(0.35, "lines"),
  point.padding = unit(0.3, "lines"), vjust = 1)
dev.off()

p <- ggplot(allmelt, aes(x = pos, y = lod, colour = pop))
png("/home/jmiller1/public_html/no_annot_.qtl.png", width = 2000)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 24) + scale_y_continuous(limits = c(0,
  22)) + scale_color_manual(values = popcol) + geom_line(size = 2, alpha = 0.85) +
  theme_minimal() + theme(axis.text = element_text(size = 10)) + labs(x = "Chromosome",
  y = "LOD", linetype = "")
dev.off()

p <- ggplot(allmelt, aes(x = pos, y = lod, colour = pop))
png("/home/jmiller1/public_html/ahr_genes_only.qtl.png", width = 2000)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 24) + scale_y_continuous(limits = c(0,
  22)) + scale_color_manual(values = popcol) + geom_line(size = 2, alpha = 0.6) +
  theme_minimal() + theme(axis.text = element_text(size = 10)) + geom_label_repel(aes(x = pos,
  y = 0.1, label = gene), box.padding = unit(0.5, "lines"), parse = T, point.padding = unit(0.8,
  "lines"), force = 1, label.padding = unit(0.2, "lines"), ylim = c(5, 20), segment.size = 1,
  max.iter = 6000, data = nbh.gens, direction = "y", size = 3, seed = 666, nudge_y = 0.01,
  vjust = 0.1) + geom_label_repel(aes(x = pos, y = 0.1, label = gene), box.padding = unit(0.5,
  "lines"), parse = T, point.padding = unit(0.8, "lines"), force = 1, label.padding = unit(0.2,
  "lines"), ylim = c(5, 20), segment.size = 0, max.iter = 6000, data = nbh.gens,
  direction = "y", size = 3, seed = 666, nudge_y = 0.01, vjust = 0.1) + labs(x = "Chromosome",
  y = "LOD", linetype = "")
dev.off()

p <- ggplot(allmelt, aes(x = pos, y = lod, colour = pop))
png("/home/jmiller1/public_html/All_chr.qtl.png", width = 3000)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 24) + scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-5, 22)) + geom_label_repel(aes(x = pos, y = 0.1,
  label = gene), color = "black", segment.size = 1, label.padding = unit(0.2, "lines"),
  data = nbh.gens, direction = "y", size = 3, seed = 666, ylim = c(5, 20), segment.alpha = 0.5) +
  geom_line(size = 2, alpha = 0.5) + geom_label_repel(aes(x = pos, y = 0.1, label = gene),
  color = "black", label.padding = unit(0.2, "lines"), ylim = c(5, 20), segment.size = 0,
  max.iter = 6000, nudge_y = 2, data = nbh.gens, direction = "y", size = 3.5, seed = 666) +
  geom_label_repel(aes(x = pos, y = 0, fill = pop, label = rank), data = all.rank,
    size = 3.5, segment.size = 1, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
      "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  geom_label_repel(aes(x = pos, y = 0, label = rank, ), fontface = "bold", data = all.rank,
    size = 3.5, segment.size = 0, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
      "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  theme(axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
    legend.position = "none") + labs(x = "Chromosome", y = "LOD", linetype = "")
dev.off()


qtlmelt <- allmelt[which(allmelt$chr %in% c( 2, 13, 18)),]
qtl.rank <- qtl.rank[which(qtl.rank$rank <21),]

qtl.rank <- all.rank[which(all.rank$chr %in% c(2, 13, 18)),]
qtl.gens <- nbh.gens[which(nbh.gens$chr %in% c( 2, 13, 18)),]
qtl.gens <- qtl.gens[as.character(c(1,3,4,5,6,7,8,24,26,45,46,56,70,96,94,57)),]
qtl.gens <- qtl.gens[c(1,3,4,5,6,7),]
qtl.gens <- qtl.gens[-5,]

p <- ggplot(qtlmelt, aes(x = pos, y = lod, colour = pop))
pdf("/home/jmiller1/public_html/all_pop_qtl_only.pdf",width=15,height=11)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 10) +
  scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-5, 22)) +
  ### gene line
  geom_label_repel(aes(x = pos, y = 0.1, label = gene),fill='white', color = "black", segment.size = 1,
    label.padding = unit(0.2, "lines"), data = qtl.gens, direction = "y", size = 5,
    seed = 666, nudge_y = 2, ylim = c(5, 20), segment.alpha = 0.5) + ## plot lines

  geom_line(size = 2, alpha = 0.75) + ### gene label

  geom_label_repel(aes(x = pos, y = 0.1, label = gene), color = "black", segment.size = 0,
    label.padding = unit(0.2, "lines"), data = qtl.gens, direction = "y", size = 5,
    seed = 666, nudge_y = 2, ylim = c(5, 20), max.iter = 6000, ) + ### Rank line

  geom_label_repel(aes(x = pos, y = 0, color = pop, label = rank), data = qtl.rank,
    size = 5, segment.size = 1, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
    "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
    ### Rank label

  geom_label_repel(aes(x = pos, y = 0, color = pop,label = rank), fontface = "bold", data = qtl.rank,
    size = 5, segment.size = 0, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
    "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +

  theme(axis.title.x = element_blank(),
    axis.title.y = element_text(face = "bold", color = "black", size = 16),
    axis.text.y=element_text(face = "bold", color = "black", size = 16),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    strip.text.x = element_text(face = "bold", color = "black",size=16),
    legend.position = "none") +
  labs(x = "Chromosome", y = "LOD", linetype = "")
dev.off()
## Why would chromosomes with top ranked loci have fewest signatures?

p <- ggplot(qtlminor, aes(x = pos, y = lod, colour = pop))
png("/home/jmiller1/public_html/minor_qtl_only.png", width = 2000)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 9) + scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-1, 8)) + theme_minimal() + geom_label_repel(aes(x = pos,
  y = 0.1, label = gene), color = "black", segment.size = 1, label.padding = unit(0.2,
  "lines"), data = minor.gens, direction = "y", size = 5, seed = 666, nudge_y = 2,
  ylim = c(5, 20), segment.alpha = 0.5) + geom_line(size = 2, alpha = 0.5) + geom_label_repel(aes(x = pos,
  y = 0.1, label = gene), color = "black", label.padding = unit(0.2, "lines"),
  ylim = c(5, 20), segment.size = 0, max.iter = 6000, nudge_y = 2, data = minor.gens,
  direction = "y", size = 5, seed = 666) + geom_label_repel(aes(x = pos, y = 0,
  color = pop, label = rank), data = minor.rank, size = 5, segment.size = 1, force = 4,
  min.segment.length = 0.1, point.padding = unit(0.4, "lines"), direction = "both",
  ylim = c(0, -8), seed = 666, box.padding = 0.1) + geom_label_repel(aes(x = pos,
  y = 0, label = rank), fontface = "bold", data = minor.rank, size = 5, segment.size = 0,
  force = 4, min.segment.length = 0.1, point.padding = unit(0.4, "lines"), direction = "both",
  ylim = c(0, -8), seed = 666, box.padding = 0.1) + theme(axis.title.x = element_blank(),
  axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
  labs(x = "Chromosome", y = "LOD", linetype = "")
dev.off()

p <- ggplot(incompat, aes(x = pos, y = lod, colour = pop))
png("/home/jmiller1/public_html/elr_incompat.png", width = 750)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 9) +
  theme(strip.text.x = element_text(size = 12)) +
  scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-1, 8)) +
  geom_label_repel(aes(x = pos, y = 0.1, label = gene), color = "black", segment.size = 1,
    label.padding = unit(0.2, "lines"), data = incompat.gens, direction = "y",
    size = 5, seed = 666, nudge_y = 2, ylim = c(5, 20), segment.alpha = 0.5) +
  geom_line(size = 2, alpha = 0.5) + geom_label_repel(aes(x = pos, y = 0.1, label = gene),
  color = "black", label.padding = unit(0.2, "lines"), ylim = c(5, 20), segment.size = 0,
  max.iter = 6000, nudge_y = 2, data = incompat.gens, direction = "y", size = 5,
  seed = 666) +
  geom_label_repel(aes(x = pos, y = 0, color = pop, label = rank),
  data = incompat.rank, size = 5, segment.size = 1, force = 4, min.segment.length = 0.1,
  point.padding = unit(0.4, "lines"), direction = "both", ylim = c(0, -8), seed = 666,
  box.padding = 0.1) +
  geom_label_repel(aes(x = pos, y = 0, label = rank), fontface = "bold",
  data = incompat.rank, size = 5, segment.size = 0, force = 4, min.segment.length = 0.1,
  point.padding = unit(0.4, "lines"), direction = "both", ylim = c(0, -8), seed = 666,
  box.padding = 0.1) + theme(axis.title.x = element_blank(), axis.text.x = element_blank(),
  axis.ticks.x = element_blank(), legend.position = "none") + labs(x = "Chromosome",
  y = "LOD", linetype = "")
dev.off()

### QTL only

p <- ggplot(ol.melt, aes(x = pos, y = lod, colour = pop))
png("/home/jmiller1/public_html/ol_qtl_only.png", width = 2000)
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 9) +
scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-1, 8)) +
  theme_minimal() + geom_label_repel(aes(x = pos,
  y = 0.1, label = gene), color = "black", segment.size = 1, label.padding = unit(0.2,
  "lines"), data = minor.gens, direction = "y", size = 5, seed = 666, nudge_y = 2,
  ylim = c(5, 20), segment.alpha = 0.5) +
  geom_line(size = 2, alpha = 0.5) +
  geom_label_repel(aes(x = pos,y = 0.1, label = gene),
  color = "black", label.padding = unit(0.2, "lines"),
  ylim = c(5, 20), segment.size = 0, max.iter = 6000, nudge_y = 2, data = minor.gens,
  direction = "y", size = 5, seed = 666) +
  geom_label_repel(aes(x = pos, y = 0,color = pop, label = rank),
  data = minor.rank, size = 5, segment.size = 1, force = 4,
  min.segment.length = 0.1, point.padding = unit(0.4, "lines"), direction = "both",
  ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  geom_label_repel(aes(x = pos,y = 0, label = rank),
  fontface = "bold", data = minor.rank, size = 5, segment.size = 0,
  force = 4, min.segment.length = 0.1, point.padding = unit(0.4, "lines"), direction = "both",
  ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  theme(axis.title.x = element_blank(),
  axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "none") +
  labs(x = "Chromosome", y = "LOD", linetype = "")
dev.off()


names(popcol)[2] <- 'BRP'

p <- ggplot(ol.melt, aes(x = pos, y = lod, colour = pop))
pdf("/home/jmiller1/public_html/OL_chr.qtl.pdf")
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 24) +
  scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-5, 22)) +
  geom_label_repel(aes(x = pos, y = 0.1,label = gene),data = ol.gens,
    color = "black", segment.size = 1, label.padding = unit(0.2, "lines"),
    direction = "y", size = 3, seed = 666, ylim = c(5, 20), segment.alpha = 0.5) +
  geom_line(size = 2, alpha = 0.5) +
  geom_label_repel(aes(x = pos, y = 0.1, label = gene), data = ol.gens,
    color = "black", label.padding = unit(0.2, "lines"), ylim = c(5, 20), segment.size = 0,
    max.iter = 6000, nudge_y = 2, direction = "y", size = 3.5, seed = 666) +
  geom_label_repel(aes(x = pos, y = 0, fill = pop, label = rank), data = ol.rank,
    size = 3.5, segment.size = 1, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
      "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  geom_label_repel(aes(x = pos, y = 0, label = rank, ), fontface = "bold", data = ol.rank,
    size = 3.5, segment.size = 0, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
      "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  theme(axis.title.y = element_text(face='bold',size = 12),
    strip.text.x = element_text(face='bold',size = 12),
    axis.text.x = element_text(face='bold', size = 12),
    axis.text.y = element_text(face='bold', size = 12),
    legend.position = "none") +
  labs(x = "Chromosome", y = "LOD", linetype = "")
dev.off()


qtl_pg <- c(1,5, 10, 12, 23)
ol.rank <- all.rank[which(all.rank$chr %in% qtl_pg), ]
ol.melt <- allmelt[which(allmelt$chr %in% qtl_pg), ]
ol.gens <- nbh.gens[which(nbh.gens$chr %in% qtl_pg), ]


p <- ggplot(ol.melt, aes(x = pos, y = lod, colour = pop))
pdf("/home/jmiller1/public_html/OL_noQTL_chr.qtl.pdf")
p + facet_wrap(~chr, nrow = 1, scales = "free_x", ncol = 24) +
  scale_color_manual(values = popcol) +
  scale_y_continuous(limits = c(-5, 22)) +
  geom_label_repel(aes(x = pos, y = 0.1,label = gene),data = ol.gens,
    color = "black", segment.size = 1, label.padding = unit(0.2, "lines"),
    direction = "y", size = 3, seed = 666, ylim = c(5, 20), segment.alpha = 0.5) +
  geom_line(size = 2, alpha = 0.5) +
  geom_label_repel(aes(x = pos, y = 0.1, label = gene), data = ol.gens,
    color = "black", label.padding = unit(0.2, "lines"), ylim = c(5, 20), segment.size = 0,
    max.iter = 6000, nudge_y = 2, direction = "y", size = 3.5, seed = 666) +
  geom_label_repel(aes(x = pos, y = 0, fill = pop, label = rank), data = ol.rank,
    size = 3.5, segment.size = 1, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
      "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  geom_label_repel(aes(x = pos, y = 0, label = rank, ), fontface = "bold", data = ol.rank,
    size = 3.5, segment.size = 0, force = 4, min.segment.length = 0.1, point.padding = unit(0.4,
      "lines"), direction = "both", ylim = c(0, -8), seed = 666, box.padding = 0.1) +
  theme(axis.title.y = element_text(face='bold',size = 12),
    strip.text.x = element_text(face='bold',size = 12),
    axis.text.x = element_text(face='bold', size = 12),
    axis.text.y = element_text(face='bold', size = 12),
    legend.position = "none") +
  labs(x = "Chromosome", y = "LOD", linetype = "")
dev.off()


### Entropy

NBH <- subset(cross.NBH, ind = cross.NBH$pheno$gt == 1)
png("/home/jmiller1/public_html/nbh_entropy.png", width = 3000)
plotInfo(NBH, chr = c(1:24), main = "NBH", method = "both", include.genofreq = T)
dev.off()

NEW <- subset(cross.NEW, ind = cross.NEW$pheno$gt == 1)
png("/home/jmiller1/public_html/new_entropy.png", width = 3000)
plotInfo(NEW, chr = c(1:24), main = "NEW", method = "both", include.genofreq = T)
dev.off()

ELR <- subset(cross.ELR, ind = cross.ELR$pheno$gt == 1)
png("/home/jmiller1/public_html/elr_entropy.png", width = 3000)
plotInfo(ELR, chr = c(1:24), main = "ELR", method = "both", include.genofreq = T)
dev.off()

BRP <- subset(brp.remap, ind = brp.remap$pheno$gt == 1)
png("/home/jmiller1/public_html/brp_entropy.png", width = 3000)
plotInfo(BRP, chr = c(1:24), main = "BRP", method = "both", include.genofreq = T)
dev.off()

#### Density

pheno <- read.table("/home/jmiller1/QTL_Map_Raw/popgen/rQTL/data/pheno.csv",
  stringsAsFactors = T, header = T,sep=',')
pheno$pop_all <- factor(pheno$pop_all, levels = rev(c('NBH','BRP','NEW','ELR')))
pheno$pheno_all <- factor(pheno$pheno_all, levels = c(NA,0:5))
pheno$gtd <- pheno$pheno_all
pheno[which(pheno$GT_NG_ALT=='NG'),6] <- NA
